OpenTelemetry proto formatter for access logging

itamarkam opened the pull request:
Signed-off-by: Itamar Kaminski <itamark@google.com>

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: A dedicated OpenTelemetryFormatter used for access logging
Additional Description: Before this change, the OpenTelemetry access logger used the "StructFormatter", so it had to convert OT format to Struct and back which wasted a lot of CPU. This change copies most of the logic from the StructFormatter and creates a formatter that uses OpenTelemetry e2e.
Risk Level: Low - only used in OT access logging, without I/O changes.
Testing:
Unit tests and integration tests
Microbenchmarks - A new microbenchamrks for the OpenTelemetry formatter was added. In order to test the difference in the access logger performance (that doesn't have a benchmark), I created an additional microbenchmark (locally) that performs the OpenTelemetry access logger's flow from before this change - Take OTel config, convert to Struct, format using StructFormatter, convert the output back to OTel.
Results show a significant difference, reducing time and CPU by ~30x!:
$ bazel run -c opt test/common/formatter:substitution_formatter_speed_test  --config=remote

Benchmark                                   Time             CPU   Iterations
--------------------------------------- --------------------------------------
BM_OldOpenTelemetryAccessLogFormatter            182017 ns       181623 ns         5446
BM_OpenTelemetryAccessLogFormatter                      6001 ns          6000 ns        98773


Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

lizan commented:
/assign-from @envoyproxy/first-pass-reviewers

rojkov commented on source/common/formatter/substitution_formatter.h:
```diff
@@ -241,6 +243,95 @@ class StructFormatter {
+  // properties preserved between the format and the log entry, thus std::map.
+  using OpenTelemetryFormatMap = std::map<std::string, OpenTelemetryFormatValue>;
+  using OpenTelemetryFormatMapPtr = std::unique_ptr<OpenTelemetryFormatMap>;
+  struct OpenTelemetryFormatMapWrapper {
+    OpenTelemetryFormatMapPtr value_;
+  };
```
A drive-be question: what's the benefit of `OpenTelemetryFormatMapWrapper` over `OpenTelemetryFormatMapPtr`?

itamarkam replied:
I just copied the whole StructFormatter as is and converted the types to match OpenTelemetry logging.
I tried templating it but it adds a lot of boiler plate which I'm not sure is worth the trouble.

itamarkam commented:
> drive-by: do you have a microbenchmark showing the improvement from this approach? It looks like you do. Can you put results into the PR description?

With the new OTel formatter, I added a microbenchmark for the formatter. The benchmark is for the formatter, not the access logger, so there's nothing to compare to, so I added a microbenchmark locally that does what the access logger used to do beforehand - take OT config, convert to Struct, format with StructFormatter, convert output back to OT.
Results show a significant difference, reducing time and CPU by ~30x! (also added to the description):
$ bazel run -c opt test/common/formatter:substitution_formatter_speed_test  --config=remote

-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_OldOpenTelemetryAccessLogFormatter            182017 ns       181623 ns         5446
BM_OpenTelemetryAccessLogFormatter                      6001 ns          6000 ns        98773

mattklein123 reviewed (dismissed):
Thanks LGTM. Should this change have a release note?

/wait

itamarkam commented:
/retest

itamarkam commented:
> @itamarkam please never force push as it makes reviews much more difficult. Thanks!

Yeah you're right, sorry for that. There were no code changes requested so I thought it wouldn't matter that much but it does make it more complicated.

itamarkam commented:
/retest

itamarkam commented:
/retest

KBaichoo commented:
/retest

itamarkam commented:
/retest
